unix/w0vncserver/PipeWireStream.cxx

```diff
@@ -0,0 +1,319 @@
+  }
+}
+
+void PipeWireStream::handleStreamStateChanged(enum pw_stream_state /* old */,
+                                              enum pw_stream_state state,
+                                              const char* error)
+{
+  switch (state) {
+  case PW_STREAM_STATE_UNCONNECTED:
+    vlog.debug("PipeWire stream disconnected");
+    break;
+  case PW_STREAM_STATE_PAUSED:
+    pw_stream_set_active(stream, true);
+    break;
+  case PW_STREAM_STATE_ERROR:
+    fatal_error("%s", error);
+    return;
+  default:
+    vlog.debug("Unhandled stream state: %d", state);
+    break;
+  }
+}
+
+void PipeWireStream::handleStreamParamChanged(uint32_t id,
+                                              const spa_pod *param)
+{
+  uint8_t paramsBuffer[1024];
+  spa_pod_builder builder;
+  int nParams;
+  const spa_pod *params[5];
+  int32_t mult;
+  spa_video_info spaFormat;
+  spa_rectangle fbSize;
+  int32_t fbStride;
+  rfb::PixelFormat pf;
+
+  if (param == nullptr || id != SPA_PARAM_Format) {
+    vlog.debug("clear format");
+    return;
+  }
+
+  if (param != nullptr && id == SPA_PARAM_Tag) {
+    spa_debug_pod(0, nullptr, param);
+    return;
+  }
+
+  if (spa_format_parse(param, &spaFormat.media_type,
+                       &spaFormat.media_subtype) < 0)
+    return;
+
+  if (spaFormat.media_type != SPA_MEDIA_TYPE_video)
+    return;
+
+  spa_pod_builder_init(&builder, paramsBuffer, sizeof(paramsBuffer));
+  spa_debug_format(2, NULL, param);
+
+  switch (spaFormat.media_subtype) {
+  case SPA_MEDIA_SUBTYPE_raw:
+    spa_format_video_raw_parse(param, &spaFormat.info.raw);
+    fbSize = SPA_RECTANGLE(spaFormat.info.raw.size.width,
+                         spaFormat.info.raw.size.height);
+
+    try {
+      pf = convertPixelformat(SPA_VIDEO_FORMAT_BGRx);
+    } catch (std::runtime_error& e) {
+      pw_stream_set_error(stream, -EINVAL, "%s", e.what());
+      return;
+    }
+
+    mult = pf.bpp / 8;
+    fbStride = spaFormat.info.raw.size.width * mult;
+    break;
+  case SPA_VIDEO_FORMAT_UNKNOWN:
+    pw_stream_set_error(stream, -EINVAL, "unknown pixel format");
+    return;
+  default:
+    pw_stream_set_error(stream, -EINVAL, "unsupported pixel format");
+    return;
+  }
+
+  if (!fbSize.width || !fbSize.height) {
+    pw_stream_set_error(stream, -EINVAL, "invalid framebuffer size: %dx%d",
+                                          fbSize.width, fbSize.height);
+    return;
+  }
+
+  setParameters(fbSize.width, fbSize.height, pf);
+
+  nParams = 0;
+
+  params[nParams++] = (spa_pod*)spa_pod_builder_add_object(&builder,
+    SPA_TYPE_OBJECT_ParamBuffers, SPA_PARAM_Buffers,
+    SPA_PARAM_BUFFERS_buffers, SPA_POD_CHOICE_RANGE_Int(4, 2, 8),
+    SPA_PARAM_BUFFERS_blocks, SPA_POD_Int(1),
+    SPA_PARAM_BUFFERS_size, SPA_POD_Int(fbSize.width * fbSize.height * mult),
+    SPA_PARAM_BUFFERS_stride, SPA_POD_Int(fbStride),
+    SPA_PARAM_BUFFERS_dataType, SPA_POD_CHOICE_FLAGS_Int((1 << SPA_DATA_MemPtr)));
```

Comment:
This should be `SPA_DATA_MemFd` instead, because it gets converted to this buffer type anyway when `PW_STREAM_FLAG_MAP_BUFFERS` is set. See https://gitlab.freedesktop.org/pipewire/pipewire/-/blob/master/src/pipewire/stream.c#L269.